include/base_interface.h

```diff
@@ -0,0 +1,322 @@
+   * matrices (i.e. as it is usally done in standard FE codes)
+   */
+  virtual void assemble_local_matrices (const typename DoFHandler<dim,spacedim>::active_cell_iterator &cell,
+                                        FEValuesCache<dim,spacedim> &scratch,
+                                        CopyData &data) const;
+
+  virtual void get_local_system_residual (const typename DoFHandler<dim,spacedim>::active_cell_iterator &cell,
+                                          FEValuesCache<dim,spacedim> &scratch,
+                                          CopyData &data) const;
+
+  /**
+   * Compute linear operators needed by the problem
+   *
+   * This function is used to assemble linear operators related
+   * to the problem.
+   */
+  virtual void compute_system_operators(const DoFHandler<dim,spacedim> &,
+                                        const std::vector<shared_ptr<typename LAC::BlockMatrix> >,
+                                        LinearOperator<typename LAC::VectorType> &,
+                                        LinearOperator<typename LAC::VectorType> &) const;
+
+
+  shared_ptr<Mapping<dim,spacedim> > get_mapping() const;
+
+  virtual unsigned int set_mapping_degree() const;
+
+  virtual UpdateFlags get_face_update_flags() const;
+
+  virtual UpdateFlags get_matrices_update_flags() const;
+
+  void fix_solution_dot_derivative(FEValuesCache<dim,spacedim> &, double) const;
+
+  void fix_solution_dot_derivative(FEValuesCache<dim,spacedim> &fe_cache, Sdouble alpha) const;
+
+
+  void fix_solution_dot_derivative(FEValuesCache<dim,spacedim> &fe_cache, SSdouble alpha) const;
+
+  template<typename Number>
+  void reinit(const Number &alpha,
+              const typename DoFHandler<dim,spacedim>::active_cell_iterator &cell,
+              FEValuesCache<dim,spacedim> &fe_cache) const;
+
+
+  template<typename Number>
+  void reinit(const Number &alpha,
+              const typename DoFHandler<dim,spacedim>::active_cell_iterator &cell,
+              const unsigned int face_no,
+              FEValuesCache<dim,spacedim> &fe_cache) const;
+
+
+
+  const Table<2,DoFTools::Coupling> &get_matrix_coupling(const unsigned int &i) const;
+  virtual unsigned int get_number_of_matrices() const;
+
+protected:
+
+  void build_couplings();
+
+  /**
+   * Define the coupling of each matrix among its blocks.  By default it
+   * sets a fully coupling among each block.  If you want to specify the
+   * coupling you need to override this function and implement it
+   * according to the following example
+   * @code
+   * void set_matrices_coupling(std::vector<std::string &couplings) const
+   * {
+   *   // suppose we have 2 matrices (system and preconditioner)
+   *   // we are solving incompressible Stokes equations
+   *   couplings[0] = "1,1;1,0"; // first block is the velocity, the second is the pressure
+   *   couplings[1] = "1,0;0,1";
+   * }
+   */
+  virtual void set_matrices_coupling(std::vector<std::string> &couplings) const;
```

Comment:
matrices here is used as an adjective: it should be singular. `set_matrix_couplings`